### ISSUE TITLE
Unsafe use of hashing for message authentication

### ISSUE THREAD

**Issue by sveeke** [Coinversable/validana-core#2]
This might not be the best place for this issue. Please move it when there is a better repo for it :).

While reviewing crypto code, we observe that it uses SHA-256 module which does not guarantee authentication as provided by HMAC. Hashing is to be used only for tamper-evidence mechanism - not as the message authentication mechanism. It is recommended to use HMAC-SHA256 and use public key as passphrase.

**Comment by wdenbakker** [comment]
I assume this is in reference to the sign/verify functions used in [key.ts](https://github.com/Coinversable/validana-core/blob/master/src/basics/key.ts).

We make use of ECDSA (Elliptic Curve Digital Signature Algorithm) for signing/verifying our transactions.
The first step in this algorithm is to hash the data you want to sign (or verify). As such the Encryption.createVerify() and Encryption.createSign() methods require a hashing algorithm. In our case: sha256. (More accurately, we use double sha256 which is why we hash the data we input once more. This due to double sha256 being the standard for blockchains, due to its resistance against length extension attacks.) The actual signing/verification is then passed a private/public key to generate/verify a signature.

In other words: the SHA256 hash we use is only for a hashing step in the ECDSA algorithm, which we use for authentication instead of HMAC.

(We use ECDSA instead of HMAC because HMAC is meant for communication between only 2 parties, where in blockchains every node will be a party, thus there being many parties.)
